From a8cdab2d11c5fad211a78b693bad8d7da3a6ae4b Mon Sep 17 00:00:00 2001
From: Geetha sowjanya <gakula@marvell.com>
Date: Tue, 10 Sep 2019 11:20:58 +0530
Subject: [PATCH 339/767] iommu/arm-smmu-v3: Force 32 byte command queue memory
 reads

commit 830d9512c07504eaee8f2066c12c4cc420ac15a7 from
git@git.assembla.com:cavium/WindRiver.linux.git
on SMMU for 96xx and 95xx silicons

64 and 128 bytes SMMU memory reads are broken
on 95xx A0/A1 and 96xx A0/B0 chips. Hence force
SMMU to issue 32 byte memory reads always.

Change-Id: I6186f56b6427970ac9a6d8b880fbc282c6a9b92e
Signed-off-by: Geetha sowjanya <gakula@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/15615
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/iommu/arm-smmu-v3.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/drivers/iommu/arm-smmu-v3.c b/drivers/iommu/arm-smmu-v3.c
index 36efeaaf782c..be788aee14e7 100644
--- a/drivers/iommu/arm-smmu-v3.c
+++ b/drivers/iommu/arm-smmu-v3.c
@@ -88,6 +88,9 @@
 
 #define ARM_SMMU_IIDR			0x18
 #define IIDR_CN96XX_A0			0x2b20034c
+#define IIDR_CN96XX_B0			0x2b20134c
+#define IIDR_CN95XX_A0			0x2b30034c
+#define IIDR_CN95XX_A1			0x2b30134c
 
 #define ARM_SMMU_CR0			0x20
 #define CR0_ATSCHK			(1 << 4)
@@ -2984,14 +2987,17 @@ static int arm_smmu_device_hw_probe(struct arm_smmu_device *smmu)
 	/* Options based on implementation */
 	reg = readl_relaxed(smmu->base + ARM_SMMU_IIDR);
 
+	/* Marvell Octeontx2 SMMU wrongly issues unsupported
+	 * 64 byte memory reads under certain conditions for
+	 * reading commands from the command queue.
+	 * Force command queue drain for every two writes,
+	 * so that SMMU issues only 32 byte reads.
+	 */
 	switch (reg) {
 	case IIDR_CN96XX_A0:
-		/* Marvell Octeontx2 SMMU wrongly issues unsupported
-		 * 64 byte memory reads under certain conditions for
-		 * reading commands from the command queue.
-		 * Force command queue drain for every two writes,
-		 * so that SMMU issues only 32 byte reads.
-		 */
+	case IIDR_CN96XX_B0:
+	case IIDR_CN95XX_A0:
+	case IIDR_CN95XX_A1:
 		smmu->options |= ARM_SMMU_OPT_FORCE_QDRAIN;
 		break;
 	}
-- 
2.31.1

